src/main/cpp/src/zorder.cu

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2020-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.cuh>
+#include <cudf/strings/detail/utilities.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+
+#include "zorder.hpp"
+
+namespace spark_rapids_jni {
+
+std::unique_ptr<cudf::column> interleave_bits(
+  cudf::table_view const& tbl,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr) {
+
+  auto num_columns = tbl.num_columns();
+  CUDF_EXPECTS(num_columns > 0, "The input table must have at least one column.");
+
+
+  CUDF_EXPECTS(is_fixed_width(tbl.begin()->type()), "Only fixed width columns can be used");
+
+  auto const type_id = tbl.begin()->type().id();
+  auto const base_size = cudf::size_of(tbl.begin()->type());
+  CUDF_EXPECTS(
+    std::all_of(tbl.begin(),
+                tbl.end(),
+                [type_id](cudf::column_view const& col) { return col.type().id() == type_id; }),
+    "All columns of the input table must be the same type column type.");
+
+  // Because the input is a table we know that they all have the same length.
+  auto num_rows = tbl.num_rows();
+
+  const cudf::size_type max_bytes_allowed = std::numeric_limits<cudf::size_type>::max();
+
+  int64_t tmp_column_data_size = num_rows * base_size;
+  if (tmp_column_data_size > max_bytes_allowed) {
```

Comment:
Yes it is needed because I am converting a column of INT to a column of binary. 2^31 INTs are allowed, but that results in 2^31 * 4 bytes, which is too large to fit.  Generally we should not allow this because we try to target only 2GiB columns at most, but if we have a batch that is too large for some reason we could run into issues.